### ISSUE TITLE
lmod/diffutils fails to build on OSX, clang-11

### ISSUE THREAD

**Issue by renderdude** [spack/spack#13484]
Upgraded to the 0.13, and latest develop. When trying to build lmod, diffutils fails to build. This does not appear to be an issue with diffutils, but how the compiler wrapper is interacting with clang-11 as I can build diffutils just fine outside fo spack.

### Steps to reproduce the issue

$ spack install diffutils

### Error Message

==> libiconv is already installed in /Users/mbolsta/spack/opt/spack/darwin-mojave-x86_64/clang-11.0.0-apple/libiconv-1.16-jg6ekuhbklculym7pyjpx6l5jetym7wp
==> Installing diffutils
==> Searching for binary cache of diffutils
==> Warning: No Spack mirrors are currently configured
==> No binary for diffutils found: installing from source
==> Using cached archive: /Users/mbolsta/spack/var/spack/cache/_source-cache/archive/b3/b3a7a6221c3dc916085f0d205abf6b8e1ba443d4dd965118da364a1dc1cb3a26.tar.xz
==> Staging archive: /var/folders/34/c56014h11_j6qr9nnwz9tphw0038d0/T/mbolsta/spack-stage/spack-stage-diffutils-3.7-yn2yzi35s3ls5cw4zjso5rswe6rivsob/diffutils-3.7.tar.xz
==> Created stage in /var/folders/34/c56014h11_j6qr9nnwz9tphw0038d0/T/mbolsta/spack-stage/spack-stage-diffutils-3.7-yn2yzi35s3ls5cw4zjso5rswe6rivsob
==> No patches needed for diffutils
==> Building diffutils [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 1:
    '/var/folders/34/c56014h11_j6qr9nnwz9tphw0038d0/T/mbolsta/spack-stage/spack-stage-diffutils-3.7-yn2yzi35s3ls5cw4zjso5rswe6rivsob/spack-src/configure' '--prefix=/Users/mbolsta/spack/opt/spack/darwin-mojave-x86_64/clang-11.0.0-apple/diffutils-3.7-yn2yzi35s3ls5cw4zjso5rswe6rivsob'

2 errors found in build log:
     136    checking for ld used by /Users/mbolsta/spack/lib/spack/env/clang/cl
            ang... /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeD
            efault.xctoolchain/usr/bin/ld
     137    checking if the linker (/Applications/Xcode.app/Contents/Developer/
            Toolchains/XcodeDefault.xctoolchain/usr/bin/ld) is GNU ld... no
     138    checking for shared library run path origin... done
     139    checking 32-bit host C ABI... no
     140    checking for the common suffixes of directories in the library sear
            ch path... lib,lib
     141    checking for libsigsegv... no, consider installing GNU libsigsegv
  >> 142    checking how /Users/mbolsta/spack/lib/spack/env/clang/clang reports
             undeclared, standard C functions... configure: error: in `/private
            /var/folders/34/c56014h11_j6qr9nnwz9tphw0038d0/T/mbolsta/spack-stag
            e/spack-stage-diffutils-3.7-yn2yzi35s3ls5cw4zjso5rswe6rivsob/spack-
            src/spack-build':
  >> 143    configure: error: cannot detect from compiler exit status or warnin
            gs
     144    See `config.log' for more details

### Information on your system

OSX 10.14.6

`--> cc --version
Apple clang version 11.0.0 (clang-1100.0.33.8)
Target: x86_64-apple-darwin18.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
-----

**Comment by renderdude** [comment]
More information:

The error above came when I tried to build lmod or diffutils from a clean install. I then clean again, bootstrapped, and built gcc, py-numpy@1.16.0, and py-pip (packages.yaml is forcing 2.7.16). When looking at the build, I notice that diffutils had been built successfully. I was then able to go back and build lmod. No idea why this worked in this round-about way

**Comment by madmissmolly** [comment]
Had same issue on Linux Ubuntu 16.04 when building new release:

```
Installing diffutils
==> Searching for binary cache of diffutils
==> Warning: No Spack mirrors are currently configured
==> No binary for diffutils found: installing from source
==> Using cached archive: /spack/var/spack/cache/_source-cache/archive/b3/b3a7a6221c3dc916085f0d205abf6b8e1ba443d4dd965118da364a1dc1cb3a26.tar.xz
==> Staging archive: /tmp/hdiags/spack-stage/spack-stage-diffutils-3.7-iywcimufvqpnga47hcct65tq2s4djrv3/diffutils-3.7.tar.xz
==> Created stage in /tmp/hdiags/spack-stage/spack-stage-diffutils-3.7-iywcimufvqpnga47hcct65tq2s4djrv3
==> No patches needed for diffutils
==> Building diffutils [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 1:
    '/tmp/hdiags/spack-stage/spack-stage-diffutils-3.7-iywcimufvqpnga47hcct65tq2s4djrv3/spack-src/configure' '--prefix=/spack/opt/spack/linux-ubuntu16.04-skylake/clang-10.0.0/diffutils-3.7-iywcimufvqpnga47hcct65tq2s4djrv3'

2 errors found in build log:
     134    checking for ld used by /spack/lib/spack/env/clang/clang... /usr/bin/ld
     135    checking if the linker (/usr/bin/ld) is GNU ld... yes
     136    checking for shared library run path origin... done
     137    checking 32-bit host C ABI... no
     138    checking for the common suffixes of directories in the library search path... lib,lib
     139    checking for libsigsegv... no, consider installing GNU libsigsegv
  >> 140    checking how /spack/lib/spack/env/clang/clang reports undeclared, standard C functions... configure: error: in `/tmp/hdiags/spack-stage/spack-stage-diffutils-3.7-iywcimufvqpnga47hcct65tq2s4dj
            rv3/spack-src/spack-build':
  >> 141    configure: error: cannot detect from compiler exit status or warnings
     142    See `config.log' for more details

See build log for details:
  /tmp/hdiags/spack-stage/spack-stage-diffutils-3.7-iywcimufvqpnga47hcct65tq2s4djrv3/spack-build-out.txt

```

**Comment by madmissmolly** [comment]
When I specify the compiler for diffutils to be clang it fails but it builds successfully with gcc:

```
$ spack install diffutils%clang
==> libiconv is already installed in /spack/opt/spack/linux-ubuntu16.04-skylake/clang-10.0.0/libiconv-1.16-ulbyhjuebct75ax4xadwegvncx5lpngv
==> Installing diffutils
==> Searching for binary cache of diffutils
==> Warning: No Spack mirrors are currently configured
==> No binary for diffutils found: installing from source
==> Using cached archive: /spack/var/spack/cache/_source-cache/archive/b3/b3a7a6221c3dc916085f0d205abf6b8e1ba443d4dd965118da364a1dc1cb3a26.tar.xz
==> Staging archive: /tmp/hdiags/spack-stage/spack-stage-diffutils-3.7-iywcimufvqpnga47hcct65tq2s4djrv3/diffutils-3.7.tar.xz
==> Created stage in /tmp/hdiags/spack-stage/spack-stage-diffutils-3.7-iywcimufvqpnga47hcct65tq2s4djrv3
==> No patches needed for diffutils
==> Building diffutils [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Error: ProcessError: Command exited with status 1:
    '/tmp/hdiags/spack-stage/spack-stage-diffutils-3.7-iywcimufvqpnga47hcct65tq2s4djrv3/spack-src/configure' '--prefix=/spack/opt/spack/linux-ubuntu16.04-skylake/clang-10.0.0/diffutils-3.7-iywcimufvqpnga47hcct65tq2s4djrv3'

2 errors found in build log:
     134    checking for ld used by /spack/lib/spack/env/clang/clang... /usr/bin/ld
     135    checking if the linker (/usr/bin/ld) is GNU ld... yes
     136    checking for shared library run path origin... done
     137    checking 32-bit host C ABI... no
     138    checking for the common suffixes of directories in the library search path... lib,lib
     139    checking for libsigsegv... no, consider installing GNU libsigsegv
  >> 140    checking how /spack/lib/spack/env/clang/clang reports undeclared, standard C functions... configure: error: in `/tmp/hdiags/spack-stage/spack-stage-diffutils-3.7-iywcimufvqpnga47hcct65tq2s4dj
            rv3/spack-src/spack-build':
  >> 141    configure: error: cannot detect from compiler exit status or warnings
     142    See `config.log' for more details

See build log for details:
  /tmp/hdiags/spack-stage/spack-stage-diffutils-3.7-iywcimufvqpnga47hcct65tq2s4djrv3/spack-build-out.txt

$ spack install diffutils%gcc
==> libiconv is already installed in /spack/opt/spack/linux-ubuntu16.04-broadwell/gcc-5.4.0/libiconv-1.16-2lu3ocrr3xhmxxowxsuglfwtzhhjcwnr
==> Installing diffutils
==> Searching for binary cache of diffutils
==> Warning: No Spack mirrors are currently configured
==> No binary for diffutils found: installing from source
==> Using cached archive: /spack/var/spack/cache/_source-cache/archive/b3/b3a7a6221c3dc916085f0d205abf6b8e1ba443d4dd965118da364a1dc1cb3a26.tar.xz
==> Staging archive: /tmp/hdiags/spack-stage/spack-stage-diffutils-3.7-kc4ior65gw573ggbfhtknp54ix4h2pdx/diffutils-3.7.tar.xz
==> Created stage in /tmp/hdiags/spack-stage/spack-stage-diffutils-3.7-kc4ior65gw573ggbfhtknp54ix4h2pdx
==> No patches needed for diffutils
==> Building diffutils [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Executing phase: 'build'
==> Executing phase: 'install'
==> Successfully installed diffutils
  Fetch: 0.00s.  Build: 20.57s.  Total: 20.57s.
[+] /spack/opt/spack/linux-ubuntu16.04-broadwell/gcc-5.4.0/diffutils-3.7-kc4ior65gw573ggbfhtknp54ix4h2pdx
```

Perhaps this is why when it was installed as a dependency of another package by @renderdude it succeeded (as it used the gcc compiler) but by default it attempts to use clang and fails?

**Comment by renderdude** [comment]
@madmissmolly: I thought the same thing, but nope. I checked the compiler flags of the installed diffutils and it's clang-11

**Comment by nazavode** [comment]
Same issue here, I confirm that `diffutils%clang@9.0.0` is still broken.

**Comment by Char-Aznable** [comment]
Why is this closed? I tried installing diffutils%clang@12.0.0 and still got this error

**Comment by Char-Aznable** [comment]
I can confirm that applying the patch from https://github.com/spack/spack/pull/13778 resolves the issue with clang-12
